### PR TITLE
SPI API: Adding  3-wire mode support

### PR DESCRIPTION
**Pull request by tbursztyka**
I had 2 options:

- keep the same size of spi_config structure and take a bit from the slave attribute
- rework the spi_config structure 

I took the 2nd option because C bit-fields is a no-go on a public API (if I had to go with C bit-fields) or parsing operation on one attribute and some in another would have been really ugly.
Also because the lines mode (single/dual/quad/octal) are not really used anywhere: such modes require some additional changes on the SPI API to be properly used (xec SPI driver tried to use it as it is but it's not exactly right, as writing a command forces the user to configure the controller on single line, holding CS, then re-configure on dual/quad/octal in order to get the command result at the right mode: it's heavy and cumbersome and limit the possibilities as well of these modes).
To really enable these modes something like #20564 should be done.

This obviously breaks struct spi_config a bit, but macros have same names, full duplex is by default (since it's 99.999% of the cases anyway) etc...

### CODE REVIEW

**Review by scottwcpg — APPROVED** [review]

**Comment by avisconti** [comment]
@tbursztyka                                                                                 
                                                                                
It seems a very good job! Before looking at it in a more detailed way I have a question: did you use hts221 sensor to test SPI 3 wire? Have you also tested other sensors?

**Comment by tbursztyka** [comment]
@avisconti  I am in the process of testing hts221 for this purpose, this is the only 3wire sensor I have at hand. But if it works for it, it should work for the others as well (knowing that other SPI controllers than STM32 will need to be updated as well).

**Comment by tbursztyka** [comment]
Looks like some sensors can do full or half duplex, such as lis2dh (another ST sensor). I guess this will need a DTS solution in order to differentiate one or the other. (currently the spi version of this driver is obviously hard-coding full duplex mode).

**Comment by avisconti** [comment]
> @avisconti I am in the process of testing hts221 for this purpose, this is the only 3wire sensor I have at hand. But if it works for it, it should work for the others as well (knowing that other SPI controllers than STM32 will need to be updated as well).

I saw that you are also implementing multi-instance, which is good! In such a casa may I ask you,  if not bothering too much, to use the common stmemsc API you can find under sensor/stmemsc ?                                                                                   
Those routines are common among many ST sensors (mostly SPI 4-wires and I2C) and would be nice to have that part tested as well. If I'm requesting too much, I can do it later.                                                                               
                                                                               
And you may also make use of the latest DT helpers like I did in #38343 (not merged yet), in particular in commit c74640bceee66fbfa87bc019b23936225d4da2bf.  Please let me know.

**Comment by avisconti** [comment]
> Looks like some sensors can do full or half duplex, such as lis2dh (another ST sensor). I guess this will need a DTS solution in order to differentiate one or the other. (currently the spi version of this driver is obviously hard-coding full duplex mode).

Yes, I agree.

**Comment by tbursztyka** [comment]
> I saw that you are also implementing multi-instance, which is good! In such a casa may I ask you, if not bothering too much, to use the common stmemsc API you can find under sensor/stmemsc ?

Sound relevant to do it now on hts221

**Comment by avisconti** [comment]
> > I saw that you are also implementing multi-instance, which is good! In such a casa may I ask you, if not bothering too much, to use the common stmemsc API you can find under sensor/stmemsc ?
> 
> Sound relevant to do it now on hts221

@tbursztyka  Yes, and I guess you should cerry-pick c74640bceee66fbfa87bc019b23936225d4da2bf because the PR is still not merged.

**Comment by scottwcpg** [comment]
@tbursztyka I checked with Microchip HW team and the MCHP XEC QMSPI controller can only support the 3 pin half-duplex mode if the QMSPI MOSI and MISO pins are externally connected together. Need to add driver logic to return an error if the driver is configured for 3-pin mode.

**Comment by tbursztyka** [comment]
> @tbursztyka I checked with Microchip HW team and the MCHP XEC QMSPI controller can only support the 3 pin half-duplex mode if the QMSPI MOSI and MISO pins are externally connected together. 

That's always the case for all SPI controller as far as I know, so that's totally fine.

> Need to add driver logic to return an error if the driver is configured for 3-pin mode.

At first yes, I'll update all drivers so they reject SPI_HALF_DUPLEX, but it would be nice that each SPI driver owner would enable the support relevantly afterwards.

**Comment by scottwcpg** [comment]
@tbursztyka Yes, once the PR is merged I will implement support in the XEC SPI driver(s).  

Somewhat off topic: Is there any chance your PR for dual/quad SPI or another will be merged?  Or has the Zephyr community decided dual/quad should be handled by a the flash driver?  If I implement a flash driver using the same QMSPI controller I will have "duplicate" drivers because the SPI driver has no support for dual/quad.  You are correct the dual/quad support I put in the XEC SPI driver is an ugly hack.

**Comment by mbolivar-nordic** [comment]
Isn't this a stable API change? It has to go through the stable API change process if so.

**Review by mbolivar-nordic — DISMISSED** [review]
> Isn't this a stable API change? It has to go through the stable API change process if so.

I would like an answer to this question.

**Comment by tbursztyka** [comment]
> > Isn't this a stable API change? It has to go through the stable API change process if so.

Yes thus why I put the DNM label. I want first to see users reaction, then once it's tested and proven to work: I'll propose the change through https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html#stable (And I'll probably push for an updated #20564 as well)

> 
> I would like an answer to this question.

time zone, anyone?  ;)

**Comment by mbolivar-nordic** [comment]
> > > Isn't this a stable API change? It has to go through the stable API change process if so.
> 
> Yes thus why I put the DNM label.

Great, thanks; I'm dropping my -1 on that understanding then.

File: drivers/sensor/hts221/hts221.c
Comment (avisconti) [inline drivers/sensor/hts221/hts221.c:248]: Here is better to use SPI_DT_SPEC_INST_GET() and, similarly, I2C_DT_SPEC_INST_GET() for i2c bus. I'm saying so because stmemsc is going to change i/f because of #38343, which is not merged yet, so, depending which one will be merged first the other should be aligned. Not sure what the best option is. If this one will be merged first I will then change my PR, otherwise you can align this one from now (and cherry-pick my commit c74640bceee66fbfa87bc019b23936225d4da2bf, which can be merged with this PR instead of mine). What do you think?

**Review by avisconti — CHANGES_REQUESTED** [review]

File: drivers/sensor/hts221/hts221.c
Comment (avisconti) [reply drivers/sensor/hts221/hts221.c:248]: Maybe @gmarull and @mbolivar can provide their feedback about how to proceed.

Comment (tbursztyka) [reply drivers/sensor/hts221/hts221.c:248]: You PR will probably land for the next release, same for this PR as it has to go through API change process. So there will be time for adapting this PR against the changes on stmemsc.

Comment (tbursztyka) [reply drivers/sensor/hts221/hts221.c]: @avisconti change was applied btw.

File: include/drivers/spi.h
Comment (mbolivar-nordic) [inline include/drivers/spi.h:264]: Is client code expected to ifdef the size of the operation value too? Maybe it's worth adding a `typedef` for this. That would avoid some duplication here too.

Just an idea.

Comment (tbursztyka) [reply include/drivers/spi.h:264]: it will not need to ifdef for the size but for using the extended modes configuration bits (if it supports both normal single line or multiple ones). At least I don't see the size as an issue.

**Comment by tbursztyka** [comment]
@avisconti  can you re-review please?

@mbolivar-nordic Could you re-review, especially the last commit I added. There is no device at the moment that will use such property in DTS, but I hope this is the best way to expose it.

**Comment by carlescufi** [comment]
@anangl could you take a look please?

File: drivers/sensor/hts221/hts221.c
Comment (avisconti) [reply drivers/sensor/hts221/hts221.c]: > @avisconti change was applied btw.

OK, I'm reviewing right now

**Review by avisconti — APPROVED** [review]
I just checked the hts221 sensor changes, and it seems a very good job! Thx!

File: dts/bindings/spi/spi-device.yaml
Comment (mbolivar-nordic) [inline dts/bindings/spi/spi-device.yaml:22]: I think this should use an enum (https://docs.zephyrproject.org/latest/guides/dts/bindings.html#enum-values) and a default (0), with a justification for why (that's the usual case for SPI controllers).

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
One minor request for the binding.

File: drivers/spi/Kconfig
Comment (anangl) [inline drivers/spi/Kconfig:34]:
```suggestion
	  enables the possibility to select the line mode (single/dual/
```

File: include/drivers/spi.h
Comment (anangl) [inline include/drivers/spi.h:240]:
```suggestion
 *     lines               [ 16 : 17 ] - MISO lines: Single/Dual/Quad/Octal.
 *     reserved            [ 18 : 31 ] - reserved for future use.
```

Comment (tbursztyka) [reply include/drivers/spi.h:240]: indeed

**Review by anangl — APPROVED** [review]

File: include/drivers/spi.h
Comment (gmarull) [inline include/drivers/spi.h:121]: nit: add `U` suffix to all unsigned numbers

File: include/drivers/spi.h
Comment (gmarull) [inline include/drivers/spi.h:110]: nit: this will just document first define, not the others. You can use

```c
/**
 * @name SPI MISO lines ....
 * @{
 */

...

/** @} */
```

File: include/drivers/spi.h
Comment (gmarull) [inline include/drivers/spi.h:238]: `@kconfig{CONFIG_SPI_EXTENDED_MODES}`

File: drivers/spi/spi_litespi.c
Comment (gmarull) [inline drivers/spi/spi_litespi.c:49]: nit: would be nice to provide some helpers, e.g.

```c
#define SPI_GET_LINES(operation) ((operation) & SPI_LINES_MASK)
```

to then do:

```c
SPI_GET_LINES(config->operation) != SPI_LINES_SINGLE) {
```

File: drivers/sensor/hts221/hts221.c
Comment (gmarull) [inline drivers/sensor/hts221/hts221.c:28]: nit: extra space after `12.5`

File: drivers/sensor/hts221/hts221_trigger.c
Comment (gmarull) [inline drivers/sensor/hts221/hts221_trigger.c:134]: may read better if:

```c 
	if (cfg->gpio_drdy.port == NULL) {
		LOG_DBG("gpio_drdy not defined in DT");
		return 0;
	}

	if (!device_is_ready(cfg->gpio_drdy.port)) {
		LOG_ERR("device %s is not ready", cfg->gpio_drdy.port->name);
		return -ENODEV;
	}
```

File: dts/bindings/spi/spi-device.yaml
Comment (gmarull) [inline dts/bindings/spi/spi-device.yaml:22]: nit for multiline description: use `description: |`

File: include/dt-bindings/spi/spi.h
Comment (gmarull) [inline include/dt-bindings/spi/spi.h:23]: only first define gets documented, see related previous code

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, thank you. I don't have any other findings except for the ones already reported by @gmarull. I will approve once those have been fixed.

**Review by nvlsianpu — APPROVED** [review]

File: dts/bindings/spi/spi-device.yaml
Comment (mbolivar-nordic) [reply dts/bindings/spi/spi-device.yaml:22]: Seconded

File: drivers/spi/spi_litespi.c
Comment (tbursztyka) [reply drivers/spi/spi_litespi.c:49]: Let's no that in #39991

File: drivers/sensor/hts221/hts221_trigger.c
Comment (tbursztyka) [reply drivers/sensor/hts221/hts221_trigger.c:134]: I just c/p what has been done in other drivers but indeed it's more logical to read it that way

File: include/dt-bindings/spi/spi.h
Comment (tbursztyka) [reply include/dt-bindings/spi/spi.h]: Actually this is on all multiple defines, I'll make a dedicated commit for it

File: dts/bindings/spi/spi-device.yaml
Comment (tbursztyka) [reply dts/bindings/spi/spi-device.yaml:22]: only if you break the line right after description:  or for all? 

I saw confusing examples in other bindings. Anyway I'll break the line.

**Review by jhedberg — APPROVED** [review]

**Comment by carlescufi** [comment]
@mbolivar-nordic please revisit

**Review by gmarull — APPROVED** [review]

**Comment by carlescufi** [comment]
@henrikbrixandersen please revisit

**Review by erwango — APPROVED** [review]

File: drivers/sensor/hts221/hts221_trigger.c
Comment (henrikbrixandersen) [inline drivers/sensor/hts221/hts221_trigger.c:132]:
```suggestion
	if (cfg->gpio_drdy.port == NULL) {
		LOG_DBG("gpio_drdy not defined in DT");
		return 0;
	}

	if (!device_is_ready(cfg->gpio_drdy.port)) {
		LOG_ERR("device %s is not ready", cfg->gpio_drdy.port->name);
		return -ENODEV;
	}
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

File: dts/bindings/spi/spi-device.yaml
Comment (mbolivar-nordic) [reply dts/bindings/spi/spi-device.yaml]: > only if you break the line right after description: or for all?

The rules and rationale for them are documented in 
https://docs.zephyrproject.org/latest/guides/dts/bindings.html#general-rules

File: drivers/sensor/hts221/hts221.h
Comment (kapi-no) [inline drivers/sensor/hts221/hts221.h:17]: @tbursztyka, it looks like you forgot to commit this header file

**Pull request merged**
